File: krakenwidget.c
#include <Windows.h>
#include <windowsx.h>
#include <CommCtrl.h>
#include <wchar.h>

#include "app.h"
#include "curve.h"
#include "krakenwidget.h"
#include "layout.h"
#include "xalloc.h"

typedef struct {
	KrakenWidget public;
	HFONT font;
	HFONT bold_font;
	HFONT big_font;
	HWND pump;
	HWND fan;

	struct {
		HWND device;
		HWND fan;
		HWND temp;
		HWND pump;
	} labels;

	struct {
		HWND device;
		HWND fan;
		HWND temp;
		HWND pump;
	} values;
} private_KrakenWidget;

#define self KrakenWidget *this
#define wnd ((Window*)this)
#define base (this->base)
#define public (*this)
#define private (*((private_KrakenWidget*)this))

#define ID_PUMP 0x1001
#define ID_FAN 0x1002


void KrakenWidget_update(self) {
	Kraken *k = public.kraken;
	wchar_t buffer[20];

	swprintf(buffer, 20, L"%.1f \u00b0C", k->temp_c);
	SetWindowText(private.values.temp, buffer);

	swprintf(buffer, 20, L"%.0f rpm", k->fan_rpm);
	SetWindowText(private.values.fan, buffer);

	swprintf(buffer, 20, L"%.0f rpm", k->pump_rpm);
	SetWindowText(private.values.pump, buffer);
}

static void position(self) {

	int m = Window_scale(wnd, 5);
	int h = Window_scale(wnd, 20);

	LayoutCell c_lab_Device = { .control = private.labels.device };
	LayoutCell c_lab_Temp =   { .control = private.labels.temp };
	LayoutCell c_lab_Fan =    { .control = private.labels.fan };
	LayoutCell c_lab_Pump =   { .control = private.labels.pump };

	LayoutCell c_val_Device = { .control = private.values.device };
	LayoutCell c_val_Temp =   { .control = private.values.temp, .text = L"99.9 \u00b0C" };
	LayoutCell c_val_Fan =    { .control = private.values.fan,  .text = L"9999 rpm", };
	LayoutCell c_val_Pump =   { .control = private.values.pump, .text = L"9999 rpm", };

	LayoutCell *cells[8] = {
		&c_lab_Device, &c_val_Device,
		&c_lab_Temp,   &c_val_Temp,
		&c_lab_Fan,    &c_val_Fan,
		&c_lab_Pump,   &c_val_Pump,
	};
	
	HDC hdc = GetDC(base.hwnd);
	Layout(hdc, 0, 0, 4, 2, cells, Window_scale(wnd, 5));

	int pump_y = c_val_Pump.y + c_val_Pump.ascender;
	int fan_y = c_val_Fan.y + c_val_Fan.ascender;
	int val_w = max(max(c_val_Pump.x + c_val_Pump.width, c_val_Fan.x + c_val_Fan.width), c_val_Temp.x + c_val_Temp.width);

	MoveWindow(private.pump, val_w + m, pump_y - h, base.width - val_w - m, h, TRUE);
	MoveWindow(private.fan, val_w + m, fan_y - h, base.width - val_w - m, h, TRUE);
}

static void load_assets(self) {
	if (private.font) DeleteObject(private.font);
	if (private.bold_font) DeleteObject(private.bold_font);
	if (private.big_font) DeleteObject(private.big_font);

	private.font = CreateFont(
		Window_scale(wnd, 16), 0, 0, 0, 0, 0, 0, 0, DEFAULT_CHARSET, 0, 0, 0, 0, L"Segoe UI");

	private.bold_font = CreateFont(
		Window_scale(wnd, 16), 0, 0, 0, FW_BOLD, 0, 0, 0, DEFAULT_CHARSET, 0, 0, 0, 0, L"Segoe UI");

	private.big_font = CreateFont(
		Window_scale(wnd, 36), 0, 0, 0, FW_BOLD, 0, 0, 0, DEFAULT_CHARSET, 0, 0, 0, 0, L"Arial");

	SetWindowFont(private.pump, private.font, FALSE);
	SetWindowFont(private.fan, private.font, FALSE);

	SetWindowFont(private.labels.device, private.font, FALSE);
	SetWindowFont(private.labels.fan, private.font, FALSE);
	SetWindowFont(private.labels.pump, private.font, FALSE);
	SetWindowFont(private.labels.temp, private.font, FALSE);

	SetWindowFont(private.values.device, private.bold_font, FALSE);
	SetWindowFont(private.values.fan, private.big_font, FALSE);
	SetWindowFont(private.values.pump, private.big_font, FALSE);
	SetWindowFont(private.values.temp, private.big_font, FALSE);

	InvalidateRect(base.hwnd, NULL, TRUE);
}

static HBRUSH static_color(self, HDC hdc, HWND ctrl) {

	if (ctrl == private.labels.device || ctrl == private.labels.fan ||
		ctrl == private.labels.pump || ctrl == private.labels.temp) {
		SetTextColor(hdc, RGB(110, 110, 110));
	} else {
		SetTextColor(hdc, RGB(30, 30, 30));
	}

	SetBkMode(hdc, TRANSPARENT);
	return GetSysColorBrush(COLOR_WINDOW);
}


static void resize(self) {
	load_assets(this);
	position(this);
}

static void created(self) {
	load_assets(this);
}

static int selected(self, int command) {

	if (command == ID_PUMP) {
		int i = ComboBox_GetCurSel(private.pump);
		if (i != CB_ERR) {
			SetCursor(LoadCursor(NULL, IDC_APPSTARTING));
			const Curve *c = Curve_pump_presets[i];
			Kraken_set_pump_curve(public.kraken, c);
		}
		return 0;
	}

	if (command == ID_FAN) {
		int i = ComboBox_GetCurSel(private.fan);
		if (i != CB_ERR) {
			SetCursor(LoadCursor(NULL, IDC_APPSTARTING));
			const Curve *c = Curve_fan_presets[i];
			Kraken_set_fan_curve(public.kraken, c);
		}
		return 0;
	}

	return 1;
}

static WindowClass class = {
	.name = L"KrakenWidget",
	.style = WS_CHILD | WS_VISIBLE,
	.created = (fn_window) created,
	.resize = (fn_window) resize,
	.select = (fn_window_command) selected,
	.static_color = (fn_window_static_color) static_color,
};

KrakenWidget *KrakenWidget_create(Window *parent, Kraken *kraken) {
	self = xmalloc(sizeof(private_KrakenWidget));
	base.class = &class;
	private.font = NULL;
	private.bold_font = NULL;
	private.big_font = NULL;
	public.kraken = kraken;

	Window_init(wnd, parent, L"");

	#define MAKE_DROPDOWN(id) CreateWindowEx( \
		0, WC_COMBOBOX, L"", CBS_DROPDOWNLIST | CBS_HASSTRINGS | WS_CHILD | WS_OVERLAPPED | WS_VISIBLE, \
		0, 0, 0, 1000, base.hwnd, (HMENU) id, App_instance, NULL)

	private.pump = MAKE_DROPDOWN(ID_PUMP);
	private.fan = MAKE_DROPDOWN(ID_FAN);
    
    wchar_t buf[128] = {0};
	for (int i = 0; Curve_pump_presets[i]; i++) {
        wcscpy_s(buf, 128, Curve_pump_presets[i]->name);
		ComboBox_AddString(private.pump, buf);
    }
    
	for (int i = 0; Curve_fan_presets[i]; i++) {
        wcscpy_s(buf, 128, Curve_fan_presets[i]->name);
		ComboBox_AddString(private.fan, buf);
    }

	#define MAKE_STATIC(text, s) CreateWindowEx( \
		0, WC_STATIC, text, s | WS_CHILD | WS_VISIBLE /*| SS_SUNKEN*/, \
		0, 0, 0, 0, base.hwnd, (HMENU) NULL, App_instance, NULL)

	private.labels.fan = MAKE_STATIC(L"Fan", SS_RIGHT);
	private.labels.pump = MAKE_STATIC(L"Pump", SS_RIGHT);
	private.labels.device = MAKE_STATIC(L"Device", SS_RIGHT);
	private.labels.temp = MAKE_STATIC(L"Temp", SS_RIGHT);

	private.values.fan = MAKE_STATIC(L"", SS_LEFT);
	private.values.pump = MAKE_STATIC(L"", SS_LEFT);
	private.values.device = MAKE_STATIC(kraken->device->serial, SS_LEFT);
	private.values.temp = MAKE_STATIC(L"", SS_LEFT);

	return this;
}

File: krakenwidget.zig
const std = @import("std");
const app = @import("app.zig");
const curves = @import("curve.zig");
const kraken_mod = @import("kraken.zig");
const layout = @import("layout.zig");
const w = @import("win32.zig");
const window = @import("window.zig");

const Kraken = kraken_mod;
const DeviceInfo = kraken_mod.DeviceInfo;
const Curve = curves.Curve;
const LayoutCell = layout.LayoutCell;

extern fn Window_init(wnd: *window.Window, parent: ?*window.Window, title: [*:0]const u16) callconv(.c) void;
extern fn Window_scale(wnd: *window.Window, dimension: i32) callconv(.c) i32;

pub const KrakenWidget = extern struct {
    base: window.Window,
    kraken: *Kraken,
};

const Controls = extern struct {
    device: ?w.HWND,
    fan: ?w.HWND,
    temp: ?w.HWND,
    pump: ?w.HWND,
};

const PrivateKrakenWidget = extern struct {
    public: KrakenWidget,
    font: ?w.HFONT,
    bold_font: ?w.HFONT,
    big_font: ?w.HFONT,
    pump: ?w.HWND,
    fan: ?w.HWND,
    labels: Controls,
    values: Controls,
};

const ID_PUMP = 0x1001;
const ID_FAN = 0x1002;

fn selfFromBase(base: *window.Window) *PrivateKrakenWidget {
    const pub_kw: *KrakenWidget = @fieldParentPtr("base", base);
    return @fieldParentPtr("public", pub_kw);
}

fn setWindowFont(hwnd: ?w.HWND, font: ?w.HFONT, redraw: bool) void {
    if (hwnd == null) return;
    const hfont = if (font) |f| @intFromPtr(f) else 0;
    _ = w.SendMessageW(hwnd.?, w.WM_SETFONT, hfont, @intFromBool(redraw));
}

fn comboBoxAddString(hwnd: w.HWND, text: [*:0]const u16) void {
    _ = w.SendMessageW(hwnd, w.CB_ADDSTRING, 0, @bitCast(@as(isize, @intCast(@intFromPtr(text)))));
}

fn comboBoxGetCurSel(hwnd: w.HWND) i32 {
    return @intCast(@as(isize, @bitCast(w.SendMessageW(hwnd, w.CB_GETCURSEL, 0, 0))));
}

fn formatLabel(comptime fmt: []const u8, args: anytype) [:0]u16 {
    var utf8_buf: [64]u8 = undefined;
    const utf8 = std.fmt.bufPrint(&utf8_buf, fmt, args) catch unreachable;
    return std.unicode.utf8ToUtf16LeAllocZ(app.allocator, utf8) catch unreachable;
}

fn updateExport(this: *KrakenWidget) callconv(.c) void {
    const info: *const DeviceInfo = kraken_mod.getInfo(this.kraken) orelse return;

    const temp = formatLabel("{d:.1} °C", .{info.temp_c});
    defer app.allocator.free(temp);
    _ = w.SetWindowTextW(selfFromBase(&this.base).values.temp.?, temp.ptr);

    const fan = formatLabel("{d:.0} rpm", .{info.fan_rpm});
    defer app.allocator.free(fan);
    _ = w.SetWindowTextW(selfFromBase(&this.base).values.fan.?, fan.ptr);

    const pump = formatLabel("{d:.0} rpm", .{info.pump_rpm});
    defer app.allocator.free(pump);
    _ = w.SetWindowTextW(selfFromBase(&this.base).values.pump.?, pump.ptr);
}

fn position(self: *PrivateKrakenWidget) void {
    if (self.pump == null or self.fan == null) return;

    const m = Window_scale(&self.public.base, 5);
    const h = Window_scale(&self.public.base, 20);

    var c_lab_device = LayoutCell{ .text = null, .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.labels.device };
    var c_lab_temp = LayoutCell{ .text = null, .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.labels.temp };
    var c_lab_fan = LayoutCell{ .text = null, .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.labels.fan };
    var c_lab_pump = LayoutCell{ .text = null, .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.labels.pump };

    var c_val_device = LayoutCell{ .text = null, .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.values.device };
    var c_val_temp = LayoutCell{ .text = @constCast(std.unicode.utf8ToUtf16LeStringLiteral("99.9 °C")), .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.values.temp };
    var c_val_fan = LayoutCell{ .text = @constCast(std.unicode.utf8ToUtf16LeStringLiteral("9999 rpm")), .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.values.fan };
    var c_val_pump = LayoutCell{ .text = @constCast(std.unicode.utf8ToUtf16LeStringLiteral("9999 rpm")), .font = null, .height = 0, .width = 0, .ascender = 0, .x = 0, .y = 0, .control = self.values.pump };

    var cells = [_]*LayoutCell{
        &c_lab_device, &c_val_device,
        &c_lab_temp, &c_val_temp,
        &c_lab_fan, &c_val_fan,
        &c_lab_pump, &c_val_pump,
    };

    const hdc = w.GetDC(self.public.base.hwnd) orelse return;
    defer _ = w.ReleaseDC(self.public.base.hwnd, hdc);

    layout.layout(hdc, 0, 0, 4, 2, &cells, Window_scale(&self.public.base, 5));

    const pump_y = c_val_pump.y + c_val_pump.ascender;
    const fan_y = c_val_fan.y + c_val_fan.ascender;
    const val_w = @max(@max(c_val_pump.x + c_val_pump.width, c_val_fan.x + c_val_fan.width), c_val_temp.x + c_val_temp.width);

    _ = w.MoveWindow(self.pump.?, val_w + m, pump_y - h, self.public.base.width - val_w - m, h, w.TRUE);
    _ = w.MoveWindow(self.fan.?, val_w + m, fan_y - h, self.public.base.width - val_w - m, h, w.TRUE);
}

fn loadAssets(self: *PrivateKrakenWidget, reload: bool) void {
    if (reload) {
        if (self.font) |font| _ = w.DeleteObject(@ptrCast(font));
        if (self.bold_font) |font| _ = w.DeleteObject(@ptrCast(font));
        if (self.big_font) |font| _ = w.DeleteObject(@ptrCast(font));
        self.font = null;
        self.bold_font = null;
        self.big_font = null;
    }

    if (self.font == null) {
        self.font = w.createFont(
            Window_scale(&self.public.base, 16),
            std.unicode.utf8ToUtf16LeStringLiteral("Segoe UI"),
            .{},
        );
    }

    if (self.bold_font == null) {
        self.bold_font = w.createFont(
            Window_scale(&self.public.base, 16),
            std.unicode.utf8ToUtf16LeStringLiteral("Segoe UI"),
            .{ .weight = w.FW_BOLD },
        );
    }

    if (self.big_font == null) {
        self.big_font = w.createFont(
            Window_scale(&self.public.base, 36),
            std.unicode.utf8ToUtf16LeStringLiteral("Arial"),
            .{ .weight = w.FW_BOLD },
        );
    }

    setWindowFont(self.pump, self.font, false);
    setWindowFont(self.fan, self.font, false);

    setWindowFont(self.labels.device, self.font, false);
    setWindowFont(self.labels.fan, self.font, false);
    setWindowFont(self.labels.pump, self.font, false);
    setWindowFont(self.labels.temp, self.font, false);

    setWindowFont(self.values.device, self.bold_font, false);
    setWindowFont(self.values.fan, self.big_font, false);
    setWindowFont(self.values.pump, self.big_font, false);
    setWindowFont(self.values.temp, self.big_font, false);

    _ = w.InvalidateRect(self.public.base.hwnd, null, w.TRUE);
}

fn staticColor(base: *window.Window, hdc: w.HDC, ctrl: w.HWND) callconv(.c) ?w.HBRUSH {
    const self = selfFromBase(base);

    if ((self.labels.device != null and ctrl == self.labels.device.?) or
        (self.labels.fan != null and ctrl == self.labels.fan.?) or
        (self.labels.pump != null and ctrl == self.labels.pump.?) or
        (self.labels.temp != null and ctrl == self.labels.temp.?)) {
        _ = w.SetTextColor(hdc, w.RGB(110, 110, 110));
    } else {
        _ = w.SetTextColor(hdc, w.RGB(30, 30, 30));
    }

    _ = w.SetBkMode(hdc, w.TRANSPARENT);
    return w.GetSysColorBrush(w.COLOR_WINDOW);
}

fn resize(base: *window.Window) callconv(.c) void {
    const self = selfFromBase(base);
    loadAssets(self, false);
    position(self);
}

fn created(base: *window.Window) callconv(.c) void {
    loadAssets(selfFromBase(base), false);
}

fn dpi(base: *window.Window) callconv(.c) void {
    loadAssets(selfFromBase(base), true);
}

fn selected(base: *window.Window, command: i32) callconv(.c) bool {
    const self = selfFromBase(base);

    if (command == ID_PUMP) {
        const i = comboBoxGetCurSel(self.pump.?);
        if (i != w.CB_ERR) {
            _ = w.SetCursor(w.LoadCursorW(null, w.IDC_APPSTARTING));
            const c = curves.pump_presets[@intCast(i)] orelse return true;
            kraken_mod.setPumpCurve(self.public.kraken, c);
        }
        return true;
    }

    if (command == ID_FAN) {
        const i = comboBoxGetCurSel(self.fan.?);
        if (i != w.CB_ERR) {
            _ = w.SetCursor(w.LoadCursorW(null, w.IDC_APPSTARTING));
            const c = curves.fan_presets[@intCast(i)] orelse return true;
            kraken_mod.setFanCurve(self.public.kraken, c);
        }
        return true;
    }

    return false;
}

var class = window.Window.WindowClass{
    .name = std.unicode.utf8ToUtf16LeStringLiteral("KrakenWidget"),
    .registered = false,
    .style = w.WS_CHILD | w.WS_VISIBLE,
    .background = null,
    .created = created,
    .paint = null,
    .static_color = staticColor,
    .destroyed = null,
    .resize = resize,
    .dpi = dpi,
    .command = null,
    .clicked = null,
    .select = selected,
    .timer = null,
};

fn makeStatic(parent: w.HWND, text: [*:0]const u16, style: u32) ?w.HWND {
    return w.createWindow(
        std.unicode.utf8ToUtf16LeStringLiteral("Static"),
        text,
        .{
            .style = style | w.WS_CHILD | w.WS_VISIBLE,
            .x = 0,
            .y = 0,
            .width = 0,
            .height = 0,
            .parent = parent,
            .instance = app.instance,
        },
    );
}

fn createExport(parent: *window.Window, kraken: *Kraken) callconv(.c) *KrakenWidget {
    const self = app.allocator.create(PrivateKrakenWidget) catch unreachable;
    self.* = .{
        .public = .{ .base = undefined, .kraken = kraken },
        .font = null,
        .bold_font = null,
        .big_font = null,
        .pump = null,
        .fan = null,
        .labels = .{ .device = null, .fan = null, .temp = null, .pump = null },
        .values = .{ .device = null, .fan = null, .temp = null, .pump = null },
    };

    self.public.base.class = &class;
    Window_init(&self.public.base, parent, std.unicode.utf8ToUtf16LeStringLiteral(""));

    self.pump = w.createWindow(
        std.unicode.utf8ToUtf16LeStringLiteral("ComboBox"),
        std.unicode.utf8ToUtf16LeStringLiteral(""),
        .{
            .style = w.CBS_DROPDOWNLIST | w.CBS_HASSTRINGS | w.WS_CHILD | w.WS_OVERLAPPED | w.WS_VISIBLE,
            .x = 0,
            .y = 0,
            .width = 0,
            .height = 1000,
            .parent = self.public.base.hwnd,
            .menu = @ptrFromInt(ID_PUMP),
            .instance = app.instance,
        },
    );

    self.fan = w.createWindow(
        std.unicode.utf8ToUtf16LeStringLiteral("ComboBox"),
        std.unicode.utf8ToUtf16LeStringLiteral(""),
        .{
            .style = w.CBS_DROPDOWNLIST | w.CBS_HASSTRINGS | w.WS_CHILD | w.WS_OVERLAPPED | w.WS_VISIBLE,
            .x = 0,
            .y = 0,
            .width = 0,
            .height = 1000,
            .parent = self.public.base.hwnd,
            .menu = @ptrFromInt(ID_FAN),
            .instance = app.instance,
        },
    );

    var i: usize = 0;
    while (curves.pump_presets[i] != null) : (i += 1) {
        comboBoxAddString(self.pump.?, curves.pump_presets[i].?.name[0..].ptr);
    }

    i = 0;
    while (curves.fan_presets[i] != null) : (i += 1) {
        comboBoxAddString(self.fan.?, curves.fan_presets[i].?.name[0..].ptr);
    }

    self.labels.fan = makeStatic(self.public.base.hwnd, std.unicode.utf8ToUtf16LeStringLiteral("Fan"), w.SS_RIGHT);
    self.labels.pump = makeStatic(self.public.base.hwnd, std.unicode.utf8ToUtf16LeStringLiteral("Pump"), w.SS_RIGHT);
    self.labels.device = makeStatic(self.public.base.hwnd, std.unicode.utf8ToUtf16LeStringLiteral("Device"), w.SS_RIGHT);
    self.labels.temp = makeStatic(self.public.base.hwnd, std.unicode.utf8ToUtf16LeStringLiteral("Temp"), w.SS_RIGHT);

    self.values.fan = makeStatic(self.public.base.hwnd, std.unicode.utf8ToUtf16LeStringLiteral(""), w.SS_LEFT);
    self.values.pump = makeStatic(self.public.base.hwnd, std.unicode.utf8ToUtf16LeStringLiteral(""), w.SS_LEFT);
    self.values.device = makeStatic(self.public.base.hwnd, kraken_mod.getIdent(kraken), w.SS_LEFT);
    self.values.temp = makeStatic(self.public.base.hwnd, std.unicode.utf8ToUtf16LeStringLiteral(""), w.SS_LEFT);

    loadAssets(self, false);
    position(self);

    return &self.public;
}

comptime {
    @export(&updateExport, .{ .name = "KrakenWidget_update", .linkage = .strong });
    @export(&createExport, .{ .name = "KrakenWidget_create", .linkage = .strong });
}
